src/analysis/convex/cone.lean

```diff
@@ -185,6 +186,159 @@ ext' $ preimage_comp.symm
+      ... = ((1 : ℝ)/2 + -((1:ℝ)/2)) • x      : by rw [add_smul]
+      ... = (0 : ℝ) • x                       : by rw [show (1:ℝ)/2 + -(1/2) = 0, by ring]
+      ... = (0 : E)                           : zero_smul ℝ x,
+  rw [this] at hkey,
+  exact h₁ hkey,
+end
+
+/-- A pointed convex cone defines a preorder. -/
+def to_preorder (S : convex_cone E) (h₁ : pointed S) : preorder E :=
+{ le := λ x y, y - x ∈ S,
+  le_refl := λ x, by change x - x ∈ S; rw [sub_self x]; exact h₁,
+  le_trans := λ x y z xy zy, by simp [(show z - x = z - y + (y - x), by abel), add_mem S zy xy] }
+
+/-- A pointed and salient cone defines a partial order. -/
+def to_partial_order (S : convex_cone E) (h₁ : pointed S) (h₂ : salient S) : partial_order E :=
+{ le_antisymm :=
+    begin
+      intros a b ab ba,
+      by_contradiction h,
+      have h' : b - a ≠ 0 := λ h'', h (eq_of_sub_eq_zero h'').symm,
+      have H := h₂ (b-a) ab h',
+      rw [neg_sub b a] at H,
+      exact H ba,
+    end,
+  ..to_preorder S h₁ }
+
+/-- A pointed and salient cone defines an `ordered_add_comm_group`. -/
+def to_ordered_add_comm_group (S : convex_cone E) (h₁ : pointed S) (h₂ : salient S) : ordered_add_comm_group E :=
+{ add_le_add_left :=
+    begin
+      intros a b hab c,
+      change c + b - (c + a) ∈ S,
+      rw [add_sub_add_left_eq_sub],
+      exact hab,
+    end,
+  ..to_partial_order S h₁ h₂,
+  ..show add_comm_group E, by apply_instance }
+
+/--
+Constructs an ordered semimodule given an `ordered_add_comm_group`, a cone, and a proof that
+the order relation is the one defined by the cone.
+-/
+def to_ordered_semimodule {α : Type*} [ordered_add_comm_group α] [semimodule ℝ α]
+  (S : convex_cone α) (h : ∀ x y : α, x ≤ y ↔ y - x ∈ S) : ordered_semimodule ℝ α :=
+{ smul_lt_smul_of_pos :=
+    begin
+      intros x y z xy hz,
+      refine lt_of_le_of_ne _ _,
+      { rw [h (z • x) (z • y), ←smul_sub z y x],
+        exact smul_mem S hz ((h x y).mp (le_of_lt xy)) },
+      { intro H,
+        have H' := congr_arg (λ r, (1/z) • r) H,
+        refine (ne_of_lt xy) _,
+        field_simp [smul_smul, div_self ((ne_of_lt hz).symm)] at H',
+        exact H' },
+    end,
+  lt_of_smul_lt_smul_of_nonneg :=
+    begin
+      intros x y z hxy hz,
+      refine lt_of_le_of_ne _ _,
+      { rw [h x y],
+        have hz' : 0 < z,
+        { refine lt_of_le_of_ne hz _,
+          rintro rfl,
+          rw [zero_smul, zero_smul] at hxy,
+          exact lt_irrefl 0 hxy },
+        have hz'' : 0 < 1/z := div_pos (by linarith) hz',
+        have hxy' := (h (z • x) (z • y)).mp (le_of_lt hxy),
+        rw [←smul_sub z y x] at hxy',
+        have hxy'' := smul_mem S hz'' hxy',
+        field_simp [smul_smul, div_self ((ne_of_lt hz').symm)] at hxy'',
+        exact hxy'' },
+      { rintro rfl,
+        exact lt_irrefl (z • x) hxy }
+    end,
+}
+
+section positive_cone
+
+variables (α : Type*) [ordered_add_comm_group α] [ordered_semimodule ℝ α]
```

Comment:
```suggestion
variables (M : Type*) [ordered_add_comm_group M] [ordered_semimodule ℝ M]
```
I think we're moving away from using `α`, `β`, etc. as names for types that have extra structure.